crates/runtime/src/datafusion.rs

```diff
@@ -1334,6 +1339,28 @@ impl DataFusion {
             }
         }
     }
+
+    /// Create or get a logical plan from the query
+    async fn get_or_create_logical_plan(
+        &self,
+        session: &SessionState,
+        sql: &str,
+    ) -> Result<LogicalPlan, DataFusionError> {
+        let cached_plans = self.cached_plans.read().await;
```

Comment:
@kczimm This is in the critical query path. We should use a lockfree data structure if possible.